src/library/scala/collection/immutable/Vector.scala

```diff
@@ -26,1284 +33,1971 @@ import scala.runtime.Statics.releaseFence
 @SerialVersionUID(3L)
 object Vector extends StrictOptimizedSeqFactory[Vector] {
 
-  def empty[A]: Vector[A] = NIL
+  def empty[A]: Vector[A] = Vector0
 
   def from[E](it: collection.IterableOnce[E]): Vector[E] =
     it match {
-      case as: ArraySeq[E] if as.length <= 32 =>
-        if (as.isEmpty) NIL
-        else {
-          val unsafeArray = as.unsafeArray
-          val len = unsafeArray.length
-          val v = new Vector(0, len, 0)
-          val display0 = new Array[Any](len)
-          if (unsafeArray.isInstanceOf[Array[AnyRef]]) {
-            System.arraycopy(unsafeArray, 0, display0, 0, len)
-          } else {
-            var i = 0
-            while (i < len) {
-              display0(i) = unsafeArray(i)
-              i += 1
-            }
-          }
-          v.display0 = display0.asInstanceOf[Array[AnyRef]]
-          v.depth = 1
-          releaseFence()
-          v
-        }
       case v: Vector[E] => v
       case _ =>
         val knownSize = it.knownSize
-
         if (knownSize == 0) empty[E]
-        else if (knownSize > 0 && knownSize <= 32) {
-          val display0 = new Array[Any](knownSize)
-          it.iterator.copyToArray(display0)
-          val v = new Vector[E](0, knownSize, 0)
-          v.depth = 1
-          v.display0 = display0.asInstanceOf[Array[AnyRef]]
-          releaseFence()
-          v
+        else if (knownSize > 0 && knownSize <= WIDTH) {
+          val a1 = new Array[Any](knownSize)
+          it.iterator.copyToArray(a1)
+          new Vector1[E](a1.asInstanceOf[Arr1])
```

Comment:
Might be worth special-casing `Iterable` and `ArraySeq` here. with ArraySeq, we can share the underlying array. With more general collection.Iterable, we can delegate to its `copyToArray` which might avoid allocating an iterator.

Reply:
This should probably be similar to `append1ifSpace`